### PR TITLE
SSE2 shuffle/unshuffle functions for large types

### PR DESCRIPTION
**Pull request by jack-pappas**
This PR is a more-focused version of #86. It includes changes to add the SSE2-accelerated shuffle/unshuffle functions for types larger than the vector size (in this case, 16 bytes). The changes also allow the SSE2-accelerated functions to be used in some cases where the input data doesn't contain an "ideal" number of elements. In those cases, the number of elements is rounded down to an acceptable value and the SSE2-accelerated functions are used (which generally covers the majority of the elements in the input buffer); the simple byte-by-byte shuffle/unshuffle functions are used for the remaining elements. This is particularly important for "large" types: due to the way block sizes are calculated (prior to these changes), data shuffled with existing/older versions of blosc won't have the correct size to utilize the new SSE2-accelerated functions -- so these modifications allow the new functions to be used with existing shuffled/compressed data.

I'm opening the PR now to get feedback, but I think additional tests need to be implemented to verify data is round-tripped correctly and is backwards-compatible with the existing code (the byte-by-byte shuffle/unshuffle for large types) before this is merged. I'll implement some tests for this ASAP, but I encourage others to test the code as well (in whatever way is appropriate for their usage of blosc) to help sniff out potential edge cases.

### CODE REVIEW

**Comment by FrancescAlted** [comment]
Thanks for decoupling this feature in a proper PR.  Any idea on why TravisCI is complaining on:

```
Test  #1: test_test_basics .................***Exception: Numerical  0.00 sec
```

?

**Comment by jack-pappas** [comment]
Yes -- it looks to be due to the change on line 887 (in my version). The case where ``typesize`` is 1 and ``nbytes`` is 7 causes the function to return 0 for the block size, which results in division by zero in ``initialize_context_compression``. I'm working to fix the bug now.

**Comment by jack-pappas** [comment]
I've fixed the broken test in test_basics. I've also added some additional type sizes to test with in #90. The new tests there all pass when run against the current ``master``; I see a test failure when running them against the code in this PR though, so I'll continue bug-hunting.

**Comment by jack-pappas** [comment]
Ok, I've fixed the bit of code that was breaking the new tests in #90, and they all pass now. I'm going to implement some additional tests (within the next few days) to verify the new code in this PR can read data shuffled/compressed by the current code and vice versa.

File: blosc/shuffle.c [hunk 1]
```diff
@@ -312,12 +393,11 @@ static void
...
 {
   size_t i, j, k;
-  size_t neblock, numof16belem;
-  __m128i xmm0[4], xmm1[4];
+  const size_t numof16belem = size / (4 * sizeof(__m128i));
 
-  neblock = size / 4;
-  numof16belem = neblock / 16;
   for (i = 0, k = 0; i < numof16belem; i++, k += 4) {
+    __m128i xmm0[4], xmm1[4];
```
Comment (FrancescAlted) [inline blosc/shuffle.c]: Is this an example of intermingled declarations and code that MSVC would have problems with, or is this supported?

**Comment by FrancescAlted** [comment]
This is looking good so far.  Could you please post some small benchmark on speed improvements for typesizes larger than 16 bytes?  I am a curious guy :)

**Comment by jack-pappas** [comment]
I can post benchmarks sometime this week. However, please wait to merge this patch - I have since found that it exposes an existing bug in blosc. I'll contact you privately to discuss the issue.

**Comment by FrancescAlted** [comment]
Ah, ok.  I will be glad to discuss the problem.

**Comment by FrancescAlted** [comment]
Hey Jack, still waiting for discussing this existing bug in c-blosc that you mentioned here time ago :)

**Comment by jack-pappas** [comment]
I've figured out the bug, and it was a subtle issue hiding in my changes, not in blosc. I'm closing this issue as it's partially superseded by #129. Once those changes are merged in, I'll submit a new PR with the remaining changes from this PR (the large type support).
